### PR TITLE
CodeMirror blob: Override browser search to trigger file search

### LINKED ISSUE #40724: CodeMirror Blob: override browser Cmd+F
Currently, users need to manually focus on the blob view to search through the contents of the blob view. It's not obvious that the focus state controls the behavior of Cmd+F. There are two alternative solutions that we could implement to address this issue:

- Auto-focus on the blob view: this has the benefit of making it still possible to use the built-in Cmd+F functionality by unfocusing the blob view, but it has the problem that it's easy to miss the focus state of the blob view and it's not obvious how to recover it.
- Override the browser Cmd+F functionality to only search via CodeMirror. This has the benefit of making it obvious how to search inside the file (via CodeMirror search), but it has the problem that it's not clear how users should trigger the built-in search functionality.

Related Slack discussions https://sourcegraph.slack.com/archives/C03PPPY5T6V/p1661248589303509?thread_ts=1661247389.478549&cid=C03PPPY5T6V

We agreed that the second option is the lesser evil so that's the approach we will try first.

- [ ] https://github.com/sourcegraph/sourcegraph/pull/40739
- [ ] Trigger CM search on Cmd+F when the search bar is focused
- [ ] (optional) Add search input in the "Files" sidebar
- [ ] (optional) Update ref panel search input to include matches from file contents

### PR DESCRIPTION
**Pull request by fkling**
Closes #40724 

Revival of #40739 with some additional changes.

File pages now override the browser's built-in search trigger. CTRL/CMD + f will trigger CodeMirror's document search instead of the browser's built-in search. Because CodeMirror doesn't necessarily render all lines of a document it's not possible to user the browser's built-int search to search inside the file.

Additional changes in this PR (which I can move in a separate PR if desired):

- Introduced a special `'Mod'` modifier that represents Command/Meta on macOS and `Ctrl` elsewhere.
- Added a utility function to set the platform value in unit tests.

## Test plan

Manual testing and unit/integration tests.

## App preview:

- [Web](https://sg-web-fkling-40724-override-cmd-f.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-fkuhmqqhoa.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by fkling** [comment]
Added @valerybugakov and @umpox for the `Shortcut` changes (since that's more of platform thing).

**Review by valerybugakov — APPROVED** [review]
> to ask whether keyboard shortcuts need to be defined in `client/shared/src/keyboardShortcuts/keyboardShortcuts.ts`? It doesn't seem useful for super specific keybindings, though I can understand if they should all be defined in a single place.

First time looking at the shortcuts configuration: [`KEYBOARD_SHORTCUTS`](https://sourcegraph.sourcegraph.com/search?q=context%3Asourcegraph+const+KEYBOARD_SHORTCUTS&patternType=standard&groupBy=repo) defined in `keyboardShortcuts.ts` is used by [`useKeyboardShortcut`](https://sourcegraph.sourcegraph.com/search?q=context%3Asourcegraph+function+useKeyboardShortcut&patternType=standard&groupBy=repo) which [used by multiple consumer applications](https://sourcegraph.sourcegraph.com/search?q=context%3Asourcegraph+useKeyboardShortcut%28%27&patternType=standard&groupBy=repo). This coupling is confusing because not all shortcuts defined in `KEYBOARD_SHORTCUTS` are available in all these apps. E.g., there's no fuzzy finder in the browser extension. Also, in the web application, [we iterate over all shortcuts to generate a help page](https://sourcegraph.sourcegraph.com/search?q=context%3Asourcegraph+Object.values%28%7B+...KEYBOARD_SHORTCUTS%2C+...LEGACY_KEYBOARD_SHORTCUTS+%7D&patternType=standard&groupBy=repo). If the browser extension adds a new shortcut to the shared constant, it will break this logic.

It would be great for packages to define shortcuts separately to keep dependencies explicit so we can add them without contemplating if it will affect other applications. E.g., `WEB_ KEYBOARD_SHORTCUTS` in `client/web`. 

What do you think, @fkling?

File: client/shared/src/react-shortcuts/Shortcut.tsx [hunk 1]
```diff
@@ -4,18 +4,47 @@ import { Key, ModifierKey } from './keys'
...
+     * Which modifiers keys need to be pressed for the handler to be triggered.
+     * 'Mod' is a special value that translates to 'Meta' (Cmd) on macOS and
+     * 'Ctrl' on other platforms.
+     */
+    held?: (ModifierKey | 'Mod')[]
+    /**
+     * DO NOT USE! Scoping the keybinding to a specific element doesn't actually
+     * work.
+     */
     node?: HTMLElement | null
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/shared/src/react-shortcuts/Shortcut.tsx:23]: Yep, that makes a lot more sense! Thank you

**Comment by fkling** [comment]
> What do you think, @fkling?

Some first thoughts (that are not set in stone, just to be clear): I'm wondering whether such a list of keyboard shortcuts (and generating help from it) is even useful.
Just take the file page for example: Overriding ctrl+f is only done on file pages. It would be nice if it showed up in the help dialog but there is no reason to include it in the help dialog on non-file pages. I imagine that eventually specific pages will have specific keybindings and therefore it would be much better to generate the help from the list of keybindings that are currently registered.
I think this approach would be useful even if we don't want to have too many individual keybindings per page.

Also the `useKeyboardShortcut` API is kind of clunky. I can understand that we need a mechanism to filter out character key shortcuts, but I believe we could also do this at a different level and more transparently (e.g. inside the `Shortcut` component for example).

My other problem with the `keyboardShortcuts.ts` file (or any other grouping really) is that it decouples the definition from the implementation. In this case I feel like this might not be a good thing. As a developer, when I see that there is supposed to be a keybinding for copying the search query, and if I wanted to enable that in a specific part in the app, I still wouldn't know which functions would need to be called to implement that feature.
I get that those keybindings are likely not meant to be reused, but then why are they grouped in a shared file?

**Comment by olafurpg** [comment]
I just tested this locally and it works as expected for both the CM blob (hijacked Cmd+F) and the old one (no hijacking). LGTM 👍🏻

**Comment by umpox** [comment]
@fkling 

> Some first thoughts (that are not set in stone, just to be clear): I'm wondering whether such a list of keyboard shortcuts (and generating help from it) is even useful.

I agree with your point that having contextual help rather than a big global list would be better. Right now we don't have that. mainly because we have very few shortcuts! We still need to have the shortcuts listed somewhere predictable until we have that though.


As for the hook and grouping. I wouldn't be opposed to having components just declare their shortcuts in their own files assuming we have a way to capture them all and display per route. The only obstacle is similar to what @valerybugakov mentioned, right now we share our shortcuts across multiple environments. Components would need to be aware of the environment they're running in (e.g. web, vscode etc) so they can be sure they're not introducing a bad shortcut. This could be solved with the hook + top level grouping, but if we have any better ideas I'm all for it

**Review by umpox — COMMENTED** [review]
I'm unsure about overriding the global search shortcut. It feels like something which could be raised in our accessibility audit, as AFAIK there would be no other way to search for other text on the page.

Have we considered:
1. Focus the blob whenever it is loaded. As this would override the normal search shortcut anyway. Users would still be able to use the default browser search functionality if they navigate away from the blob. We'll likely do this anyway when we have better blob keyboard navigation.
2. Add global `f` shortcut for triggering the CM search from anywhere (similar to what Google CS does). We would need to ensure this can be disabled though, as it's a character key shortcut.

Although I'm not 100% on what the actual a11y requirements are around this. The only thing in the WCAG spec seemed more related to search forms more than anything else. I'll do a bit of digging but this is non-blocking, just there's a possibility we might need to revisit this for the accessibility audit.

**Comment by olafurpg** [comment]
@umpox The original issue https://github.com/sourcegraph/sourcegraph/issues/40724#issue-1347670894 includes a small summary of why we went with "global shortcut" instead of "trigger shortcut only when the blob view is focused". The primary concern was that it's easy to loose the focus on the blob view (click on the symbol sidebar, trigger fuzzy finder, ...) and it's not clear that you need to manually click on the blob view to search through the contents of the file.

**Comment by olafurpg** [comment]
>  it's not clear that you need to manually click on the blob view to search through the contents of the file.

I think that this concern can be somewhat addressed by adding a (token-based) cursor. However, we want to make the CodeMirror blob the default before experimenting with a cursor. The cursor should not be a blocking issue for enabling the new blob view.

**Comment by umpox** [comment]
@olafurpg Thanks for the context, yeah that makes sense to me.

Ideally we should also return focus to the blob view as much as possible (e.g. when clicking file/symbol in sidebar, closing fuzzy finder, etc). That will help alleviate this issue a lot, but agree this shouldn't block.

**Comment by fkling** [comment]
@umpox I think at this point we are still trying things out see what works and what doesn't. Thank you for looking more into this from the a11y side.

I like the idea of returning focus back to the file component, though unsure how that would really look like in practice.

Right now (and I don't actually know why/where this happens), when you press excape on the file view page focus goes back to the search input.

**Comment by olafurpg** [comment]
A few changes we can make to alleviate the pain of hijacking Cmd+F:

- add button in the blob search row to "Find with browser" (exact formulation up for debate) that triggers the built-in search. We could collect metrics on how frequently this is used to approximate the false-positive ratio from hijacking Cmd+F (the user wanted the browser's builtin Cmd+F instead of blob search)
- add search input to the files sidebar. This would be an improvement over our current Cmd+F because it adds the ability to search through collapsed directory entries
- make the ref panel search input filter more than files alone, also filter the repo and the text contents. This would be an improvement over Cmd+F because it could search through collapsed items.

**Comment by fkling** [comment]
@umpox 

> Components would need to be aware of the environment they're running in (e.g. web, vscode etc) so they can be sure they're not introducing a bad shortcut.

**Comment by olafurpg** [comment]
Alternatively, we could try to extend the blob search input to additionally search through the files sidebar and references panel. I don't know if this is technically possible, or if it's confusing from a UX perspective, but it would make it a suitable replacement for the browser's Cmd+F feature (while adding the ability to search through collapsed/non-visible items)

**Comment by valerybugakov** [comment]
> As for the hook and grouping. I wouldn't be opposed to having components just declare their shortcuts in their own files assuming we have a way to capture them all and display per route.

@umpox, yeah, I agree that it would be ideal. I was thinking about it in the constraints of the current implementation and how it's used in [other client applications](https://sourcegraph.sourcegraph.com/search?q=context%3Asourcegraph+useKeyboardShortcut%28%27&patternType=standard&groupBy=repo). If we have time to implement something better than `useKeyboardShortcut` that would enable contextual help and colocation of shortcuts + functionality, I'm up for it!

**Comment by fkling** [comment]
@valerybugakov @umpox 

>  If we have time to implement something better than `useKeyboardShortcut` that would enable contextual help and colocation of shortcuts + functionality, I'm up for it!

No sure about coolocation, but contextual help should be relatively easy I think: All shortcuts are already registered with the `ShortcutManager`. So all we really have to do is add a way to get the list of currently registered shortcuts. The help text could also be passed to `Shortcut` and registered with the manager.

Right now `useKeyboardShortcut`'s only actual feature seems to be enabling or disabling shortcuts based on the character key setting. That's also something that could be potentially moved to `Shortcut` or `ShortcutManager`.

**Pull request merged**